src/v/kafka/server/quota_manager.cc

```diff
@@ -510,100 +497,59 @@ ss::future<clock::duration> quota_manager::throttle_fetch_tp(
     co_return capped_delay;
 }
 
+const std::optional<quota_manager::global_map_t>&
+quota_manager::get_global_map_for_testing() const {
+    return _global_map;
+}
+
 // erase inactive tracked quotas. windows are considered inactive if
 // they have not received any updates in ten window's worth of time.
 void quota_manager::gc() {
-    vassert(
-      ss::this_shard_id() == _client_quotas.shard_id(),
-      "gc should only be performed on the owner shard");
     auto full_window = _default_num_windows() * _default_window_width();
     auto expire_threshold = clock::now() - 10 * full_window;
     ssx::background
       = ssx::spawn_with_gate_then(_gate, [this, expire_threshold]() {
-            return do_gc(expire_threshold);
```

Comment:
Note to reviewers: previously we would hold a shard-0-local `ssx::mutex` during the `ssx::sharded_ptr<>::update()` function, the lock of the `sharded_ptr`. Guarding the updates with a mutex is no longer necessary because all updates to the global map are done without any scheduling points, so we can rely on seastar's guarantee that none of these "global map mutating code blocks" will run concurrently.

Reply:
> Possibly we should actually add some maybe_yields in the two gc methods as I could imagine that for large maps those will lead to reactor stalls at which point we would need the mutex again?

Makes sense, thanks, I've added back the lock now and added `ss::maybe_yield()` to the map iterators. We now have a `maybe_yield` on each iterator step - I am wondering if that should be less frequent (eg. only yield `if i%100 == 0` or similar), but looking around the codebase I haven't seen this pattern being used.

> Best would obviously be if we could somehow make the gc not lazy. Could we tie it to the connection context or something?
I think that's difficult because we want to expire the trackers some seconds after we haven't seen the client, so if a client disconnected and reconnected we would expect to have its quota still. One option would be to trigger a call to `quota_manager::gc()` on every new connection (perhaps with some rate limiting) but that seems more complex than what we have right now.

Reply:
> but looking around the codebase I haven't seen this pattern being used

It's a bit witchcraft, we have `ssx::async_for_each` for this but it won't work in this loop because of the erase.

The problem is that the maybe_yield actually adds lots of overhead but it's hard to avoid.

Reply:
See https://github.com/redpanda-data/redpanda/pull/16512#issuecomment-1955028310 for background.

Reply:
Makes sense, thanks for the info!